Models

nkelava opened the pull request:
This PR will add entity models, basic seed and routes for them.

Fixes #3 and #8.

nkelava committed:
Add models and migrations

nkelava committed:
Add seed

nkelava committed:
Add routes with basic get route

bdeak4 commented on backend/src/database/seeders/20220713115148-seed-users.js:
```diff
@@ -0,0 +1,16 @@
+      {
+        email: 'admin@rasporedar.com',
+        password: 'admin',
+        createdAt: new Date(Date.now()),
```
`new Date(Date.now())` and `new Date()` do the same thing so latter is preffered

nkelava replied:
Fixed ❤️

bdeak4 reviewed (changes requested):
PR looks good to me, just fix suggestion for getting current date in seed

nkelava committed:
Remove unnecessary date argument

bdeak4 reviewed (dismissed):
Looks good to me! Ready to merge in main

abasic commented on backend/src/models/classroom.ts:
```diff
@@ -0,0 +1,57 @@
+import { DataTypes, Model, Optional } from 'sequelize';
+import sequelize from '../database/connection';
+
+interface ClassroomAttributes {
+  id: number;
+  name: string;
+  capacity: number;
+  createdAt: Date;
+  updatedAt: Date;
+}
+
+type ClassroomCreationalAttributes = Optional<ClassroomAttributes, 'id' | 'createdAt' | 'updatedAt'>;
+
+class Classroom extends Model<ClassroomAttributes, ClassroomCreationalAttributes> implements ClassroomAttributes {
+  declare id: number;
+  declare name: string;
+  declare capacity: number;
+  declare createdAt: Date;
+  declare updatedAt: Date;
+}
```
Looking at the Sequelize [documentation](https://sequelize.org/docs/v6/other-topics/typescript/#usage) there is a less verbose way of writing this that seems to be the recommended way of doing it since v6.14.0. I can not link directly to it but it is the second block of code on the page. Have you looked into it as an option?

abasic commented on backend/src/routes/classroom.ts:
```diff
@@ -11,8 +12,10 @@ const router = Router();
+router.get('/', async (req: Request, res: Response) => {
+  const classrooms = await Classroom.findAll();
+
+  res.status(200).send(classrooms);
```
There is a [`res.json`](https://expressjs.com/en/4x/api.html#res.json) method that is useful when returning JSON payload since it applies the correct `Content-Type` header and applies `JSON.stringify` to the payload.

Also, there should be no need to specify the 200 when sending a response. It should be the default.

nkelava replied:
Thanks, I've changed it.

abasic commented on backend/src/models/classroom.ts:
```diff
@@ -0,0 +1,57 @@
+import { DataTypes, Model, Optional } from 'sequelize';
+import sequelize from '../database/connection';
```
In other places, relative paths are avoided. But here (and in a couple of other ones) they are not. Is there a specific reason why it is like that?

nkelava replied:
There is no specific reason, auto import did it and I didn't recheck it.

nkelava committed:
Modify user model field because of language consistency

abasic commented on backend/src/database/migrations/20220714064701-rename-users-column-name.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Generally speaking, if a migration did not end up in the main branch, it can be edited in the pull request where it originated from. This is also valid and it can be a matter of preference and use case.

nkelava replied:
Noted 👍

abasic reviewed (changes requested):
@nk38104 @bdeak4 
First of all, I like the different approach to how the models are handled 👍 .
The couple of comments I wrote are general notes or minor code stuff.

But I would like to discuss live the actual structure of the database since I am having a harder time mapping our previous discussions and my understanding of the app with the structure implied through these models.

abasic commented on backend/src/database/migrations/20220708084442-create-user.js:
```diff
@@ -0,0 +1,42 @@
+        type: Sequelize.STRING,
+        allowNull: false,
+      },
+      form: {
+        type: Sequelize.JSONB,
+        allowNull: true,
+      },
+      raspored: {
+        type: Sequelize.JSONB,
+        allowNull: true,
+      },
```
Please remove these two fields for now and add equivalents when you deem appropriate.

nkelava replied:
Fields have been removed. Will do.

nkelava committed:
Remove user columns, add model contraints, refactor code

removed unnecessary user columns and updated migrations, add unique constraints to model fields and chaged router responses

bdeak4 reviewed (dismissed):
Looks good to me! 🚀

abasic commented on backend/src/database/migrations/20220714075227-remove-from-users-columns-timetable-form.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
```
Please use a [transaction](https://sequelize.org/docs/v6/other-topics/transactions/) in this migration (both up and down).

nkelava replied:
Added transaction to migration as requested 😄

nkelava committed:
Add transaction to migration
